Issue19 a11y

burge-is opened the pull request:
Issue https://github.com/FormidableLabs/react-live/issues/19

Adds a component to the demo code to handle keyboard accessibility.

The component:
- wraps its children in a tabbable div
- removes child tabIndexes
- handles entering and leaving the child components
- shows and hides a `role='alert'` tooltip when focused to announce new keyboard nav shortcuts (enter/esc)

burge-is commented:
@philpl @kentcdodds here's that PR

kentcdodds commented:
I played with the demo and the interaction is great! One thing is it's unclear when I'm focused on the tabgate, so it's hard to know where the keyboard focus is. Normally the browser will give an outline to the focused element. Could we make sure there's some styling for `:focus`? Or do we need to do that ourselves as consumers?

![hover](https://user-images.githubusercontent.com/1500684/28003279-c5ad471c-64f9-11e7-9ed2-fd9e01188639.gif)

burge-is commented:
Huh. Thought I saw the outline in mine. I'll explicitly add that style so there's zero question.

burge-is commented:
@kentcdodds The code is now adding a more explicitly styled outline `onFocus` and removing it `onBlur`

kentcdodds commented:
Cool. Here's what it looks like for me:

![outline](https://user-images.githubusercontent.com/1500684/28034914-e7f2263c-656f-11e7-9add-ba3e84e6c67c.gif)

So by adding the margin and padding on focus it's making things kinda jump around a little bit. I'd also rather this be done with regular CSS rather than inline styles and JavaScript events.

In addition, the blue outline is sub-optimal IMO. I wonder if there's another way we could highlight that it's focused...

---

Ok, I did a little bit of digging and just realized two things:

1. The `TabGate` component is only implemented in the demo. Haha, I thought this would be part of react-live's own implementation. Not a huge deal for me because I can copy/paste just fine, but I think this would be useful to other people
2. The reason the normal outline wasn't being applied is because `StyledProvider` had `overflow: hidden` which was hiding the outline. Kinda weird, but if you remove that then you don't need the custom logic with `onFocus` and `onBlur` _and_ you don't need to apply your own outline. Win win win!

From here, I can copy the `TabGate` for my own use, or we can figure out a way to make react-live work out of the box. That's up to @philpl :smile:

kentcdodds commented:
Oh, and if you make the changes I suggest (remove `outline: hidden`), it'll look like this: 

![working](https://user-images.githubusercontent.com/1500684/28035513-1d75774e-6572-11e7-9c36-7c074141cdbe.gif)

Much better :smile:

burge-is commented:
Awwww yaaa! I'm a huge fan of not doing the javascript-y css-y outline I added.  Happy to see it cleared up. Deleting code is always more satisfying than adding it in my book.  

I didn't look too much at the styled components, great find.  

Happy to see it working better, it is satisfying to see in action.  The gifs are such a nice way to show off page interaction.

philpl commented:
`outline` seems like a great addition and looks great :) The only problem I currently have with this is that it's only part of the example, while it's good default behaviour that the editor should include.

We could just include the TabGate in the editor but I was thinking (see my comments in the original issue) why can't we implement this just using `preventDefault` code that's part of the editor?

I mean we could include the TabGate code in the editor and call it a day, but why do we need this: `findFirstFocusableChild` and other code here?

@alexlande has made a modification a while ago where the editor's tab-capturing behaviour can be disabled: https://github.com/FormidableLabs/react-live/blob/5677482c301275e0756751a11ae43d5a96267451/src/components/Editor/index.js#L105

We can just add an enter handler there in an `onFocus` when `ignoreTabKey` is set and then change a flag in the state. When the flag is set we add the outline. Also when the flag is set and the tab key is pressed it is captured. When the enter key wasn't pressed we remove the `preventDefault` and have the ignore behaviour that the prop enables right now. This would then make tab unfocus (blur) the editor.

I think that's a much more elegant solution and we're really close to it already :)

Thanks for your effort already :) Looking great!

kentcdodds commented:
Yup, I'm definitely more interested/excited about `react-live` having this baked in. I don't care as much how it's implemented, but what @philpl says makes sense :+1:

burge-is commented:
@philpl A lot of the design decisions were made with desire of avoiding touching existing code and behaviors.  For this reason, code like the `findFocusableChild` exist purely to be a generic, reusable means to an end.  I understand your preference and am currently investigating adding the code directly to the Editor without the overhead.

One thing to watch for `onFocus` is triggered by both mouse and keyboard.  It seems wrong to make the mouse behavior now have an additional step to editing the code.  Some `eventListeners` looking for mouse clicks should do the trick there, though. They could implement the same code that is used for hitting the Enter key.  

Keyboard:   tab to Editor -> hit 'Enter' -> trigger whatever happens on activating the editor for change
Mouse:          click Editor -> trigger whatever happens on activating the editor for change

Both would reset whatever flag `onBlur`

Also because the editor is the active element we need to address what happens when the user types keys after focus. I assume this flag we'd store in state seems like it would drive two different branches of interactions when keys are pressed. One for the 'focused and waiting to enter' state and the other for the 'focused and editing' state.  

Instead of introducing a flag would it make sense to use the `contentEditable` flag for a lot of this behavior? Flipping it on and off accordingly.

Edge cases are the trouble makers in these kind of scenarios.

philpl commented:
> I understand your preference and am currently investigating adding the code directly to the Editor without the overhead.

Cheers! :tada: :smile:

> One thing to watch for onFocus is triggered by both mouse and keyboard. It seems wrong to make the mouse behavior now have an additional step to editing the code.

So we'd have to make sure that `onClick` skips the enter guard, i.e sets it to true immediately. Makes sense :+1:

> Also because the editor is the active element we need to address what happens when the user types keys after focus

It should probably skip the enter guard as well. I'd assume that if the user starts typing, that he's intentionally typing to edit the code.

> Instead of introducing a flag would it make sense to use the contentEditable flag for a lot of this behavior? Flipping it on and off accordingly.

Well, then the behaviour before is tricky. If we manage to reproduce all of this with just the `preventDefault` for tab at the right time, that'd be the cleanest approach imho. I might be wrong of course ^^

burge-is commented:
@philpl moved the logic over to where you directed, does this look more like what you were looking for?

For styling of the outline I introduced some css and added the following logic:
- `onFocus` adds class of `tabGuarded`
- `onBlur` or once editing begin removes class of `tabGuarded`
The demo won't reflect this unless it uses this updated css.

For navigation:
- Tabbing moves between elements until editing beings.
- Editing starts on the press of any key while focused on the editor. (per your great suggestion in the comment above this one)
- This 'editing state' is ended `onBlur` or if the 'Esc' key is pressed.
- The 'editing state' is handled by an internal flag named `guarded`, it is in charge of handling tab capture and triggering whether or not the outline should appear.

All in all there is now an `onFocus`, `onBlur`, Escape key handler, and special logic at the beginning of the `onKeyDown` to control toggling back and forth between `guarded` and not

kentcdodds commented:
Could you provide a gif? I use [LICECap](https://www.cockos.com/licecap/) 😄

kentcdodds commented on react-live.css:
```diff
@@ -92,3 +92,9 @@
   color: red;
 }
+
+.tabGuarded{
```
I don't think that this CSS is needed. The user agent styles should be good enough right? If it's not showing up, perhaps it's that `overflow: hidden` issue I mentioned earlier...

burge-is replied:
The problem is that now we are directly focusing the editor so it will have an outline the entire duration of the editing.... or that's what I was seeing.  I'll mess with the overflow you mentioned.

burge-is replied:
https://github.com/burge-is/react-live/blob/24be04a17906ff8641d1df29cc0f4c00dce24717/react-live.css#L13
^ this line make the Editor have no outline.  The TabGate code from before had a wrapper so this piece of CSS wasn't active.

burge-is replied:
There is an `outline: hidden` being applied to the `.prism-code` that wasn't in play with the old `TabGate` code. The `TabGate` had a wrapper and that is where we were outlining before, now it's directly on the `Editor` with it's existing css.

burge-is replied:
Looked into it further and deleting the `overflow: hidden` didn't work for me.  It's the change in implementation of previously having a wrapper but now dealing directly with the `Editor`.  Could be wrong though, but it's what I am seeing.

kentcdodds replied:
Interesting. Thanks for digging. Let's see what @philpl has to say :smile:

philpl replied:
The outline code is fine. The user should be able to deal with the details of how it should look :)

But can we put it at the top of the file, rename it to `.tab-guarded`, and make the rule more specific please (i.e. `.prism-code.tab-guarded`)? Cheers :)

philpl replied:
Btw it'll also need to be added here: https://raw.githubusercontent.com/FormidableLabs/react-live/master/src/constants/css.js

Sorry for that not being in one file right now :(

kentcdodds commented on src/constants/css.js:
```diff
@@ -93,4 +93,10 @@ export default `
   color: red;
 }
+
+.tabGuarded{
```
Some comment here

burge-is commented:
![a11y-fix](https://user-images.githubusercontent.com/6237202/28077636-4ea914c4-6630-11e7-99d1-eaf1f7caefd3.gif)

kentcdodds commented:
Nice! Thanks for sharing that! Yeah, I think that there's the `overflow: hidden` issue which is why you had to do the extra CSS stuff. If you remove that in the demo code you shouldn't have to do the extra CSS stuff for `.tabGate` :smile:

burge-is commented:
I kind of like the addition of the class for the purposes of styling or possible adding in the ARIA tips that were talked about in the issue thread.  Provides some context by way of class.  

But if removing that fixes it all again, it's another one of those win wins you mentioned and I can delete some lines of code.  Checking now.

Also, thanks for the software tip.  I really like this.  Now to find what I see some streamers using to display the keys they are pressing on screen.

kentcdodds commented:
You might be interested in [this](https://github.com/kentcdodds/ama/issues/107) :smile:

philpl commented on react-live.css:
```diff
@@ -92,3 +92,9 @@
 }
+
+.tabGuarded{
+    outline: 1px dotted #212121;
```
Can we instead make this solid? I don't think `dotted` looks rather native.

philpl commented on src/components/Editor/index.js:
```diff
@@ -102,10 +102,21 @@ class Editor extends Component {
-      document.execCommand('insertHTML', false, '&#009')
-      evt.preventDefault()
-    } else if (evt.keyCode === 13) { // Enter Key
+      if(!this.tabGuarded){
```
You can just wrap this into the above `if` statement.

Also we can remove the `ignoreTabKey` prop now, as the new behaviour is supposed to replace it.

philpl commented on src/components/Editor/index.js:
```diff
@@ -102,10 +102,21 @@ class Editor extends Component {
+      }
+    }else if (evt.keyCode && evt.keyCode === 27) {//Esc Key
+      this.tabGuarded=true
+      this.ref.classList.add("tabGuarded")
```
Can we just replace the `ref` usage and the `tabGuarded` flag with a flag inside the component state? In this case we don't need to be clever about how to handle the element, and can just use the lifecycle instead :)

philpl commented on src/components/Editor/index.js:
```diff
@@ -102,10 +102,21 @@ class Editor extends Component {
+        evt.preventDefault()
+      }
+    }else if (evt.keyCode && evt.keyCode === 27) {//Esc Key
+      this.tabGuarded=true
```
There's some stylistic divergence here (`=` instead of ` = `). Can you fix this please? I promise I'll add prettier to this project soon ;)

philpl replied:
Also when you move the flag over to the state you might want to add a check for `!tabGuarded` here

philpl commented on src/components/Editor/index.js:
```diff
@@ -170,9 +181,22 @@ class Editor extends Component {
+  onBlur = evt => {
+    this.tabGuarded =true
+    this.ref.classList.remove("tabGuarded")
+    return evt
```
This and the above doesn't need to return anything ;)

But we need to proxy both callbacks to potential props like the existing ones do:

```js
if (this.props.onKeyDown) {
  this.props.onKeyDown(evt)
}
```

burge-is commented:
@philpl When I originally added the flag I put it in the state.  However, this got tricky during `onBlur` because the render triggered by the state change caused the element we were tabbing away from to regain focus.  

I considered adding code for when should the state actually render / managing focus upon render but it felt more mentally taxing to consume than the idea of dealing with classList / a boolean flag on the object.  

Toying with it again this A.M. a bit to see what it would look like if I can get it to behave within the lifecycle.

burge-is commented:
Moved `tabGuarded` into state and got around that tab issue I was seeing.

philpl commented:
@burge-is amazing! I'll review it again once I'm at my desk and merge it. Don't worry about the difference in coding style. I'll just add prettier later :+1: thanks again!

kentcdodds commented:
Thanks so much for working on this @burge-is! This is going to be great for keyboard-only users. a11y FTW!

burge-is commented:
Thanks for tweeting it out Kent, my first go at contributing OSS and I learned a lot from you two and the process.  Keep up the good work on y'alls end.

philpl commented on src/components/Editor/index.js:
```diff
@@ -102,10 +103,17 @@ class Editor extends Component {
+    }else if (evt.keyCode === 27) {// Esc Key
+      this.setState({tabGuarded:true})
+    }else if(!(evt.shiftKey || evt.ctrlKey || evt.altKey)){
+      this.setState({tabGuarded:false})
```
Hm, okay. I see what prompted you to use the flag before. This state change triggers early so that the selection save (`this.selection = ...`) in the keyup handler isn't respected on enter.

So now when the enter key is pressed the state changes, the component updates, and the selection is harmed.

We can try an easy fix. First of all we can add the flags to the if statement `&& this.state.tabGuarded) {` etc

Then we can maybe move this logic into the keyup handler. It has to be after the rest of the logic there. ^^

Makes sense?

philpl commented:
Almost there now! Sorry, I found a bug with the enter key not respecting the saved selection 😢

burge-is commented:
Pesky render.  The logic for changing state is now at the end of `keyUp` to prevent accidentally skipping logic.  Good find.

philpl commented on src/components/Editor/index.js:
```diff
@@ -160,16 +164,35 @@ class Editor extends Component {
+
+    if (evt.keyCode === 27 && !this.state.tabGuarded) {// Esc Key
+      this.setState({tabGuarded:true})
+    }else if(!(evt.shiftKey || evt.keyCode === 27 || evt.keyCode === 9) && this.state.tabGuarded){
```
In `keyup` the `shift` key (19) triggers separately when released. So here you'll need to add `evt.keyCode === 16` as well. Just discovered that :/

Also, can we do `this.state.tabGuarded && !evt.shiftKey && evt.keyCode !== 27 /* ... */` instead? i.e. negate the group there to remove it, and put the `tabGuarded` first? Should help readability a little.

burge-is replied:
Definitely, I'll be a little more thorough with the key clicking this go round after the change too.

philpl replied:
✨ **Accessibility** ✨ — it's tough

burge-is replied:
Is the behavior your'e seeing a shift press is being treated as hitting any old key (the assume desire to edit on any key press logic)? That would be expected based on how I was thinking of it, the evt.shiftKey was originally put in to handle shift tabbing not necessarily an independent shift press.

burge-is replied:
Which I'm thinking shift+tab may be broken in the newest implementation not gonna lie I expect this out of event handlers.  They have that wonderful ability to get tangled up like earbuds if you aren't careful with your approach.  Kind of why I originally wrapped the whole thing in a blocking div.  The past trauma of dealing with edge cases was giving me flashbacks.  I'll dig a little bit further later this evening.

philpl replied:
@burge-is Yep, exactly, shift+tab is broken

philpl replied:
I just realised that we're being a bit silly here haha 😆 We can actually just set `tabGuarded` to false inside an `onChange` handler. So we create a new callback method, proxy it up and:

```js
onChange = evt => {
  if (this.props.onChange) {
    this.props.onChange(evt)
  }

  if (this.state.tabGuarded) {
    this.setState({ tabGuarded:false })
  }
}
```

That should work, right?

burge-is replied:
Sooooo much prettier.  That's a much better way to conceptualize it. I couldn't see the forest for the trees.   It reads more like English and that always feels nice. Gets us out of that keycode "magic number" hell 🔥👿🔥   

Abstractions are a developer's best friend.

burge-is commented:
Not quite there, I wrote it inline and tested with good results and was hasty with commiting.  When I placed the function on the component up with the other `on(X)` handlers I am seeing different behavior, back to the lab.

philpl commented:
@burge-is I think I'm not seeing any bugs anymore so it should be good to go? The only thing is, that I'm not seeing the highlighting / outline in the demo. I thought this must've worked sometime before, given that Kent posted a gif with it above?

burge-is commented:
Hmmm... my `onChange `function isn't being entered.  Are you not seeing that same behavior?

As far as the outline you need to make sure wherever you are testing is getting the latest css, when checking in demo I had to go manually replace that minified css file.  I just switched over to trying it out on the storybook code so that I could more rapidly check my changes and I am seeing the outline on focus.

I corrected the margin hoppy-ness while looking it over. As well as shifted logic away from `onFocus` into `onBlur` so there wasn't a need to worry about mouse vs keyboard focus (mouse was showing the outline for a second and then removing it)

But the `onChange` is what has me hesitant to say it's ready for merge.  You mentioned not seeing any problems but my latest only removes the tabGuard `onClick` and not on any change like we were aiming for

burge-is commented:
I made the last check in `keyUp ` see if the event were anything other than `Esc `or `Tab `while `tabGuarded`. This triggers in the fashion we were expecting, I saw no response when using `onChange`

So, funny to see my original over-engineered wrapper next to these small changes.  Coding is growing .... or in this case shrinking.

philpl commented:
@burge-is I'll review it this evening :smile: looking really solid now!

Yea, oftentimes the optimal solution ends up being simpler than any initial approach.... If one's lucky

burge-is commented:
Closing so that the commit history added to react-live isn't riddled with my subtle changes.
New PR is https://github.com/FormidableLabs/react-live/pull/23

kentcdodds commented:
In the future, you shouldn't need to worry about it, maintainers can just choose the ["Squash and Merge"](https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits) feature which will squash all changes into a single commit. It's pretty rad.

Alternatively, you can cleanup a PR yourself using [an interactive rebase](https://egghead.io/lessons/javascript-how-to-squash-multiple-git-commits) :smile:

philpl commented:
Yea, don't worry about it :smile: But it's not an issue, let's continue with the new PR

burge-is commented:
@kentcdodds very good to know, thank you.
